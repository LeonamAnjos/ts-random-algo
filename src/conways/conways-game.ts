/*
Conwayâ€™s Game of Life
The game is played on a grid of cells.
Each cell can be populated or unpopulated.

The game takes place over a series of generations. Each generation is derived from the previous generation by applying a set of rules as follows:

For a cell that is populated:
If it has one or less populated neighbours, it dies
If it has four or more populated neighbours, it dies
If it has  two or three populated neighbours, it survives

For a cell that is 'empty' or 'unpopulated'
If it has three neighbours, it becomes populated.

NB: All rules are applied in parallel. That is, the death of a cell in this generation does not affect its neighbours in this generation
*/

export class Cell {
  static at(x: number, y: number): Cell {
    return new Cell(x, y);
  }

  static from(str: string): Cell {
    const [x, y] = str.split(",");
    return new Cell(+x, +y);
  }

  constructor(public readonly x: number, public readonly y: number) {}

  public get neighbours(): Cell[] {
    return [
      Cell.at(this.x, this.y + 1),
      Cell.at(this.x, this.y - 1),
      Cell.at(this.x + 1, this.y),
      Cell.at(this.x - 1, this.y),
      Cell.at(this.x + 1, this.y + 1),
      Cell.at(this.x + 1, this.y - 1),
      Cell.at(this.x - 1, this.y + 1),
      Cell.at(this.x - 1, this.y - 1),
    ];
  }

  public neighboursInBound(min: Cell, max: Cell) {
    return this.neighbours
      .filter(({ x, y }) => x >= min.x && y >= min.y)
      .filter(({ x, y }) => x <= max.x && y <= max.y);
  }

  public format(): string {
    return `${this.x},${this.y}`;
  }
}

export class ConwayGame {
  private readonly min!: Cell;
  private readonly max!: Cell;

  constructor(public readonly width: number, public readonly height: number) {
    if (this.width < 1 || this.height < 1) {
      throw Error("Invalid arguments");
    }

    this.min = new Cell(1, 1);
    this.max = new Cell(width, height);
  }

  public play(populatedCells: Cell[]): Cell[] {
    if (populatedCells.length < 3) return [];

    const populated = new Set(populatedCells.map((c) => c.format()));
    const neighboursCount = new Map<string, number>();
    const survivors: Cell[] = [];

    for (const cell of populatedCells) {
      const neighbours = cell.neighbours
        .filter((c) => !this.isOutOfBound(c))
        .map((c) => c.format());

      let count = 0;
      for (const n of neighbours) {
        if (populated.has(n)) count++;

        neighboursCount.set(n, (neighboursCount.get(n) ?? 0) + 1);
      }

      if (count > 1 && count < 4) survivors.push(cell);
    }

    for (const [k, v] of neighboursCount) {
      if (v === 3 && !populated.has(k)) {
        survivors.push(Cell.from(k));
      }
    }

    return survivors.filter((c) => !!c);
  }

  private isOutOfBound({ x, y }: Cell): boolean {
    return x < this.min.x || y < this.min.y || x > this.max.x || y > this.max.y;
  }
}
