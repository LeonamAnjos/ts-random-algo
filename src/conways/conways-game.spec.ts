import { ConwayGame, Cell } from "./conways-game";

describe("Position", () => {
  describe(".neighbours", () => {
    it(`SHOULD return WHEN Position(2, 2)`, () => {
      const actual = new Cell(2, 2).neighbours;
      expect(actual).toMatchSnapshot();
    });

    it(`SHOULD return WHEN Position(1, 1)`, () => {
      const actual = new Cell(1, 1).neighbours;
      expect(actual).toMatchSnapshot();
    });
  });

  describe(".neighboursInBound", () => {
    const minPos = new Cell(1, 1);
    const maxPos = new Cell(3, 3);

    it(`SHOULD return WHEN Position(1, 1)`, () => {
      const actual = new Cell(1, 1).neighboursInBound(minPos, maxPos);
      expect(actual).toMatchSnapshot();
    });

    it(`SHOULD return WHEN Position(2, 2)`, () => {
      const actual = new Cell(2, 2).neighboursInBound(minPos, maxPos);
      expect(actual).toMatchSnapshot();
    });

    it(`SHOULD return WHEN Position(1, 2)`, () => {
      const actual = new Cell(1, 2).neighboursInBound(minPos, maxPos);
      expect(actual).toMatchSnapshot();
    });

    it(`SHOULD return WHEN Position(2, 3)`, () => {
      const actual = new Cell(2, 3).neighboursInBound(minPos, maxPos);
      expect(actual).toMatchSnapshot();
    });

    it(`SHOULD return WHEN Position(3, 1)`, () => {
      const actual = new Cell(3, 1).neighboursInBound(minPos, maxPos);
      expect(actual).toMatchSnapshot();
    });

    it(`SHOULD return WHEN Position(1, 3)`, () => {
      const actual = new Cell(1, 3).neighboursInBound(minPos, maxPos);
      expect(actual).toMatchSnapshot();
    });

    it(`SHOULD return WHEN Position(3, 3)`, () => {
      const actual = new Cell(3, 3).neighboursInBound(minPos, maxPos);
      expect(actual).toMatchSnapshot();
    });

    it(`SHOULD return WHEN Position(4, 3)`, () => {
      const actual = new Cell(4, 3).neighboursInBound(minPos, new Cell(5, 6));
      expect(actual).toMatchSnapshot();
    });
  });
});

describe("ConwayGame", () => {
  it("SHOULD create", () => {
    const conwayGame = new ConwayGame(5, 7);

    expect(conwayGame).toBeDefined();
    expect(conwayGame.width).toBe(5);
    expect(conwayGame.height).toBe(7);
  });

  it("SHOULD fail WHEN width OR hight is invalid", () => {
    expect(() => new ConwayGame(-1, 1)).toThrowError("Invalid arguments");
    expect(() => new ConwayGame(0, 1)).toThrowError("Invalid arguments");
    expect(() => new ConwayGame(1, -1)).toThrowError("Invalid arguments");
    expect(() => new ConwayGame(1, 0)).toThrowError("Invalid arguments");
  });

  describe(".play", () => {
    const conwayGame = new ConwayGame(5, 6);

    [
      [],
      [new Cell(1, 1)],
      [new Cell(2, 2)],
      [new Cell(1, 3)],
      [new Cell(1, 1), new Cell(1, 2)],
      [new Cell(2, 1), new Cell(1, 3)],
      [new Cell(1, 1), new Cell(2, 2), new Cell(4, 3)],
    ].forEach((input: Cell[], index: number) => {
      it(`SHOULD return empty populated cells #${index}`, () => {
        expect(conwayGame.play(input).length).toBe(0);
      });
    });

    [
      [Cell.at(1, 1), Cell.at(1, 2), Cell.at(1, 3)],
      [
        Cell.at(2, 4),
        Cell.at(2, 5),
        Cell.at(3, 2),
        Cell.at(3, 4),
        Cell.at(4, 3),
        Cell.at(4, 4),
      ],
    ].forEach((input, index: number) => {
      it(`SHOULD return populated cells #${index}`, () => {
        const actual = conwayGame.play(input);
        expect(actual).toMatchSnapshot();
      });
    });
  });
});
