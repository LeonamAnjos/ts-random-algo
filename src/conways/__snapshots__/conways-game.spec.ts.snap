// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConwayGame .play SHOULD return populated cells #0 1`] = `
Array [
  Cell {
    "x": 1,
    "y": 2,
  },
  Cell {
    "x": 2,
    "y": 2,
  },
]
`;

exports[`ConwayGame .play SHOULD return populated cells #1 1`] = `
Array [
  Cell {
    "x": 2,
    "y": 4,
  },
  Cell {
    "x": 2,
    "y": 5,
  },
  Cell {
    "x": 4,
    "y": 3,
  },
  Cell {
    "x": 4,
    "y": 4,
  },
  Cell {
    "x": 2,
    "y": 3,
  },
]
`;

exports[`Position .neighbours SHOULD return WHEN Position(1, 1) 1`] = `
Array [
  Cell {
    "x": 1,
    "y": 2,
  },
  Cell {
    "x": 1,
    "y": 0,
  },
  Cell {
    "x": 2,
    "y": 1,
  },
  Cell {
    "x": 0,
    "y": 1,
  },
  Cell {
    "x": 2,
    "y": 2,
  },
  Cell {
    "x": 2,
    "y": 0,
  },
  Cell {
    "x": 0,
    "y": 2,
  },
  Cell {
    "x": 0,
    "y": 0,
  },
]
`;

exports[`Position .neighbours SHOULD return WHEN Position(2, 2) 1`] = `
Array [
  Cell {
    "x": 2,
    "y": 3,
  },
  Cell {
    "x": 2,
    "y": 1,
  },
  Cell {
    "x": 3,
    "y": 2,
  },
  Cell {
    "x": 1,
    "y": 2,
  },
  Cell {
    "x": 3,
    "y": 3,
  },
  Cell {
    "x": 3,
    "y": 1,
  },
  Cell {
    "x": 1,
    "y": 3,
  },
  Cell {
    "x": 1,
    "y": 1,
  },
]
`;

exports[`Position .neighboursInBound SHOULD return WHEN Position(1, 1) 1`] = `
Array [
  Cell {
    "x": 1,
    "y": 2,
  },
  Cell {
    "x": 2,
    "y": 1,
  },
  Cell {
    "x": 2,
    "y": 2,
  },
]
`;

exports[`Position .neighboursInBound SHOULD return WHEN Position(1, 2) 1`] = `
Array [
  Cell {
    "x": 1,
    "y": 3,
  },
  Cell {
    "x": 1,
    "y": 1,
  },
  Cell {
    "x": 2,
    "y": 2,
  },
  Cell {
    "x": 2,
    "y": 3,
  },
  Cell {
    "x": 2,
    "y": 1,
  },
]
`;

exports[`Position .neighboursInBound SHOULD return WHEN Position(1, 3) 1`] = `
Array [
  Cell {
    "x": 1,
    "y": 2,
  },
  Cell {
    "x": 2,
    "y": 3,
  },
  Cell {
    "x": 2,
    "y": 2,
  },
]
`;

exports[`Position .neighboursInBound SHOULD return WHEN Position(2, 2) 1`] = `
Array [
  Cell {
    "x": 2,
    "y": 3,
  },
  Cell {
    "x": 2,
    "y": 1,
  },
  Cell {
    "x": 3,
    "y": 2,
  },
  Cell {
    "x": 1,
    "y": 2,
  },
  Cell {
    "x": 3,
    "y": 3,
  },
  Cell {
    "x": 3,
    "y": 1,
  },
  Cell {
    "x": 1,
    "y": 3,
  },
  Cell {
    "x": 1,
    "y": 1,
  },
]
`;

exports[`Position .neighboursInBound SHOULD return WHEN Position(2, 3) 1`] = `
Array [
  Cell {
    "x": 2,
    "y": 2,
  },
  Cell {
    "x": 3,
    "y": 3,
  },
  Cell {
    "x": 1,
    "y": 3,
  },
  Cell {
    "x": 3,
    "y": 2,
  },
  Cell {
    "x": 1,
    "y": 2,
  },
]
`;

exports[`Position .neighboursInBound SHOULD return WHEN Position(3, 1) 1`] = `
Array [
  Cell {
    "x": 3,
    "y": 2,
  },
  Cell {
    "x": 2,
    "y": 1,
  },
  Cell {
    "x": 2,
    "y": 2,
  },
]
`;

exports[`Position .neighboursInBound SHOULD return WHEN Position(3, 3) 1`] = `
Array [
  Cell {
    "x": 3,
    "y": 2,
  },
  Cell {
    "x": 2,
    "y": 3,
  },
  Cell {
    "x": 2,
    "y": 2,
  },
]
`;

exports[`Position .neighboursInBound SHOULD return WHEN Position(4, 3) 1`] = `
Array [
  Cell {
    "x": 4,
    "y": 4,
  },
  Cell {
    "x": 4,
    "y": 2,
  },
  Cell {
    "x": 5,
    "y": 3,
  },
  Cell {
    "x": 3,
    "y": 3,
  },
  Cell {
    "x": 5,
    "y": 4,
  },
  Cell {
    "x": 5,
    "y": 2,
  },
  Cell {
    "x": 3,
    "y": 4,
  },
  Cell {
    "x": 3,
    "y": 2,
  },
]
`;
